File: src/host/acp/questions.ts
import type * as acp from '@agentclientprotocol/sdk';
import type { Question, QuestionAnswers, QuestionBlock, QuestionOption } from '@shared/transcript';
import { activityOf, type NormalizeState } from './normalize';
import type { GrokAnswers, GrokQuestionRequest, GrokQuestionResponse } from './grokQuestions';
import { t } from '../i18n';

// The agent's own tool input, as seen in its tool_call: the form schema flattens it (Devin puts the option label into `const` and the
// description into `title`; Kimi drops the question text into the elicitation `message`), so the tool input is the richer source for labels
export interface RawQuestion {
  header?: string;
  question?: string;
  options?: { label: string; description?: string }[];
}

type Pending =
  | { kind: 'form'; blockId: string; schema: acp.ElicitationSchema; questions: Question[]; nodeId?: string; resolve: (r: acp.CreateElicitationResponse) => void }
  | { kind: 'grok'; blockId: string; questions: Question[]; nodeId?: string; resolve: (r: GrokQuestionResponse) => void };

export interface QuestionGateDeps {
  // Which transcript a request belongs to: the root session id → the root state; a child peer session id → that node's
  // state; bump marks the owning node dirty when a card mutates its transcript
  stateFor: (sessionId: string | undefined) => { state: NormalizeState; nodeId?: string; bump?: () => void } | undefined;
  // Every transcript that may hold cards (root + all child nodes): sweeps run over all of them;
  // nodeId pins a card to its owner — undefined is the root — so a reopened child's fresh `q-1`
  // can never settle the old `q-1` still sitting in the root history
  states: () => { state: NormalizeState; nodeId?: string; bump?: () => void }[];
  touch: () => void;
  log: (line: string) => void;
}

// Remembered tool inputs: enough for the questions in flight, a handful is plenty
const RAW_KEEP = 8;

// Holds the question cards waiting for the user. Devin / Kimi ask through elicitation/create (mode form), Grok through its own request;
// both become one QuestionBlock in the current agent turn, and the webview answers by block id. A resolved block stays as the record
export class QuestionGate {
  private pending = new Map<string, Pending>();
  private seq = 0;
  private raw: { toolCallId: string; questions: RawQuestion[] }[] = [];

  constructor(private deps: QuestionGateDeps) {}

  has(blockId: string): boolean { return this.pending.has(blockId); }

  get waiting(): boolean { return this.pending.size > 0; }

  // A tool_call / tool_call_update whose rawInput carries `questions` is the agent's ask-user-question tool; keep it for label enrichment
  rememberToolInput(u: acp.ToolCall | acp.ToolCallUpdate) {
    const raw = u.rawInput as { questions?: unknown } | null | undefined;
    if (!raw || !Array.isArray(raw.questions)) return;
    const questions = raw.questions.flatMap((q): RawQuestion[] => {
      const item = q as Record<string, unknown> | null;
      if (!item || typeof item !== 'object') return [];
      const options = Array.isArray(item.options)
        ? item.options.flatMap(o => {
          const opt = o as Record<string, unknown> | null;
          return opt && typeof opt.label === 'string' ? [{ label: opt.label, ...(typeof opt.description === 'string' ? { description: opt.description } : {}) }] : [];
        })
        : undefined;
      return [{
        ...(typeof item.header === 'string' ? { header: item.header } : {}),
        ...(typeof item.question === 'string' ? { question: item.question } : {}),
        ...(options ? { options } : {}),
      }];
    });
    this.raw = [...this.raw.filter(r => r.toolCallId !== u.toolCallId), { toolCallId: u.toolCallId, questions }].slice(-RAW_KEEP);
  }

  // Devin sends no toolCallId with the form: fall back to the latest remembered input with the same number of questions
  private rawFor(toolCallId: string | null | undefined, count: number): RawQuestion[] | undefined {
    if (toolCallId) return this.raw.find(r => r.toolCallId === toolCallId)?.questions;
    for (let i = this.raw.length - 1; i >= 0; i--) if (this.raw[i]!.questions.length === count) return this.raw[i]!.questions;
    return undefined;
  }

  async onElicitation(req: acp.CreateElicitationRequest, signal: AbortSignal): Promise<acp.CreateElicitationResponse> {
    if (signal.aborted) return { action: 'cancel' };
    // The request union ends in a catch-all variant, so the form fields are read through a plain shape rather than narrowed on `mode`
    const { requestedSchema: schema, toolCallId, sessionId } = req as { requestedSchema?: acp.ElicitationSchema; toolCallId?: string | null; sessionId?: string };
    if (req.mode !== 'form' || !schema) return { action: 'decline' };
    const ref = this.deps.stateFor(sessionId);
    if (!ref) {
      this.deps.log(`elicitation request for unknown session ${sessionId ?? '(request scope)'}`);
      return { action: 'cancel' };
    }
    const questions = formQuestions(schema, req.message, req._meta, this.rawFor(toolCallId, Object.keys(schema.properties ?? {}).length));
    if (!questions.length) return { action: 'decline' };
    const block = this.open(ref.state, questions, toolCallId ?? undefined, spareMessage(req.message, questions));
    ref.bump?.();
    return new Promise(resolve => {
      this.pending.set(block.id, { kind: 'form', blockId: block.id, schema, questions, nodeId: ref.nodeId, resolve });
      this.arm(block, signal);
    });
  }

  async onGrokQuestion(req: GrokQuestionRequest, signal: AbortSignal): Promise<GrokQuestionResponse> {
    if (signal.aborted) return { outcome: 'skip_interview' };
    const questions = req.questions.map((q): Question => ({
      id: q.question, text: q.question, kind: q.multiSelect ? 'multiple' : 'single',
      options: q.options.map(o => ({ id: o.label, label: o.label, ...(o.description && o.description !== o.label ? { description: o.description } : {}) })),
      other: true,
    }));
    if (!questions.length) return { outcome: 'accepted', answers: {} };
    const ref = this.deps.stateFor(req.sessionId);
    if (!ref) {
      this.deps.log(`question request for unknown session ${req.sessionId}`);
      return { outcome: 'skip_interview' };
    }
    const block = this.open(ref.state, questions, req.toolCallId);
    ref.bump?.();
    return new Promise(resolve => {
      this.pending.set(block.id, { kind: 'grok', blockId: block.id, questions, nodeId: ref.nodeId, resolve });
      this.arm(block, signal);
    });
  }

  // The webview closed the card. Only answered questions travel; skip tells the agent to go on with what it has (Grok's skip_interview,
  // a form's decline when nothing was answered — with partial answers a form is accepted as far as it goes, which is how Devin and Kimi read a partial form anyway)
  resolve(blockId: string, answers: QuestionAnswers, skip = false) {
    const p = this.pending.get(blockId);
    if (!p) return;
    const given = cleanAnswers(p.questions, answers);
    const empty = Object.keys(given).length === 0;
    this.pending.delete(blockId);
    this.settle(p, skip || empty ? 'skipped' : 'answered', given);
    if (p.kind === 'grok') {
      p.resolve(skip || empty ? { outcome: 'skip_interview', ...(empty ? {} : { partial_answers: given as GrokAnswers }) } : { outcome: 'accepted', answers: given as GrokAnswers });
    } else {
      p.resolve(empty ? { action: 'decline' } : { action: 'accept', content: formContent(p.schema, p.questions, given) });
    }
    this.deps.touch();
  }

  // The turn is over (cancelled, failed, closed): every open card is withdrawn and the agent told so
  cancelAll() {
    for (const p of this.pending.values()) {
      this.settle(p, 'cancelled');
      if (p.kind === 'grok') p.resolve({ outcome: 'skip_interview' }); else p.resolve({ action: 'cancel' });
    }
    this.pending.clear();
  }

  // One subagent was cancelled: its open cards close as cancelled and the agent is told so
  cancelFor(nodeId: string) {
    for (const p of [...this.pending.values()]) {
      if (p.nodeId !== nodeId) continue;
      this.pending.delete(p.blockId);
      this.settle(p, 'cancelled');
      if (p.kind === 'grok') p.resolve({ outcome: 'skip_interview' }); else p.resolve({ action: 'cancel' });
    }
  }

  private open(state: NormalizeState, questions: Question[], toolCallId?: string, message?: string): QuestionBlock {
    const block: QuestionBlock = { type: 'question', id: `q-${++this.seq}`, ...(toolCallId ? { toolCallId } : {}), ...(message ? { message } : {}), questions };
    const last = state.turns[state.turns.length - 1];
    if (last?.role === 'agent') { last.blocks.push(block); last.activity = activityOf(state.turns); }
    return block;
  }

  // The agent withdrew its request (session/cancel, a timeout on its side): the card closes as cancelled
  private arm(block: QuestionBlock, signal: AbortSignal) {
    signal.addEventListener('abort', () => {
      const p = this.pending.get(block.id);
      if (!p) return;
      this.pending.delete(block.id);
      this.settle(p, 'cancelled');
      if (p.kind === 'grok') p.resolve({ outcome: 'skip_interview' }); else p.resolve({ action: 'cancel' });
      this.deps.touch();
    }, { once: true });
    this.deps.touch();
  }

  private settle(p: Pending, outcome: QuestionBlock['outcome'], answers?: QuestionAnswers) {
    for (const e of this.deps.states()) {
      if (e.nodeId !== p.nodeId) continue;
      for (let i = e.state.turns.length - 1; i >= 0; i--) {
        const turn = e.state.turns[i];
        if (turn?.role !== 'agent') continue;
        const b = turn.blocks.find(b => b.type === 'question' && b.id === p.blockId);
        if (!b || b.type !== 'question') continue;
        b.outcome = outcome;
        if (answers && Object.keys(answers).length) b.answers = answers;
        turn.activity = activityOf(e.state.turns);
        e.bump?.();
        return;
      }
    }
  }
}

type Prop = Record<string, unknown>;

// One question per form property, in schema order. Labels come from the agent's tool input when it lines up with the schema
// (same count, and header / question text agreeing where both exist); the schema alone still yields a usable question
export function formQuestions(schema: acp.ElicitationSchema, message: string, meta: Record<string, unknown> | null | undefined, raw?: RawQuestion[]): Question[] {
  const props = schema.properties ?? {};
  const keys = Object.keys(props);
  const required = new Set(schema.required ?? []);
  const allowOther = meta?.['cognition.ai/allowOther'] === true;
  // Kimi joins the question texts into the message, one per line
  const lines = message.split('\n').map(l => l.trim()).filter(Boolean);
  const aligned = raw && raw.length === keys.length ? raw : undefined;
  return keys.flatMap((key, i) => {
    const prop = props[key] as Prop | undefined;
    if (!prop) return [];
    const rq = aligned?.[i];
    const usable = rq && (!str(prop.title) || !rq.header || prop.title === rq.header) && (!str(prop.description) || !rq.question || prop.description === rq.question) ? rq : undefined;
    const title = str(prop.title) ?? usable?.header;
    const text = str(prop.description) ?? usable?.question ?? (lines.length === keys.length ? lines[i] : undefined) ?? (keys.length === 1 ? str(message) : undefined) ?? title ?? key;
    const q: Question = { id: key, ...(title && title !== text ? { title } : {}), text, kind: 'single', options: [], ...(required.has(key) ? { required: true } : {}) };
    const items = prop.type === 'array' && prop.items && typeof prop.items === 'object' ? prop.items as Prop : undefined;
    if (prop.type === 'boolean') {
      q.options = [{ id: 'true', label: t('question.yes') }, { id: 'false', label: t('question.no') }];
    } else if (items) {
      q.kind = 'multiple';
      q.options = Array.isArray(items.anyOf) ? titled(items.anyOf, usable) : Array.isArray(items.enum) ? plain(items.enum) : [];
    } else if (Array.isArray(prop.oneOf)) {
      q.options = titled(prop.oneOf, usable);
    } else if (Array.isArray(prop.enum)) {
      q.options = plain(prop.enum);
    } else {
      q.kind = 'text';
      if (prop.type === 'number' || prop.type === 'integer') q.numeric = true;
    }
    if (q.kind !== 'text' && prop.type !== 'boolean' && (allowOther || (prop._meta as Prop | undefined)?.['cognition.ai/allowOther'] === true)) q.other = true;
    if (q.kind !== 'text' && q.options.length === 0) q.kind = 'text';
    return [q];
  });
}

function str(v: unknown): string | undefined {
  return typeof v === 'string' && v.trim() ? v : undefined;
}

function plain(values: unknown[]): QuestionOption[] {
  return values.flatMap(v => (typeof v === 'string' || typeof v === 'number' ? [{ id: String(v), label: String(v) }] : []));
}

// Titled choices ({ const, title, description }); a tool-input option with the same label as `const` supplies the label / description the agent meant
function titled(entries: unknown[], rq?: RawQuestion): QuestionOption[] {
  return entries.flatMap(e => {
    const o = e as Prop | null;
    if (!o || typeof o.const !== 'string') return [];
    const ro = rq?.options?.find(x => x.label === o.const);
    const label = ro?.label ?? str(o.title) ?? o.const;
    const description = ro ? ro.description : str(o.description);
    return [{ id: o.const, label, ...(description && description !== label ? { description } : {}) }];
  });
}

// The form's message earns a line of its own only when it says something the questions don't already
export function spareMessage(message: string, questions: Question[]): string | undefined {
  const m = message.trim();
  if (!m) return undefined;
  const texts = questions.map(q => q.text.trim());
  if (texts.includes(m) || texts.join('\n') === m.split('\n').map(l => l.trim()).filter(Boolean).join('\n')) return undefined;
  return m;
}

// Keep only answers to known questions with something in them; a single-select answer is one string, a multi-select a non-empty list
function cleanAnswers(questions: Question[], answers: QuestionAnswers): QuestionAnswers {
  const out: QuestionAnswers = {};
  for (const q of questions) {
    const a = answers[q.id];
    if (a === undefined) continue;
    if (q.kind === 'multiple') {
      const list = (Array.isArray(a) ? a : [a]).map(s => s.trim()).filter(Boolean);
      if (list.length) out[q.id] = list;
    } else {
      const s = (Array.isArray(a) ? a.join(', ') : a).trim();
      if (s) out[q.id] = s;
    }
  }
  return out;
}

// Answers go back typed the way the schema declared them: booleans and numbers are converted, lists stay lists, everything else is a string
function formContent(schema: acp.ElicitationSchema, questions: Question[], answers: QuestionAnswers): Record<string, acp.ElicitationContentValue> {
  const content: Record<string, acp.ElicitationContentValue> = {};
  for (const q of questions) {
    const a = answers[q.id];
    if (a === undefined) continue;
    const prop = schema.properties?.[q.id] as Prop | undefined;
    if (Array.isArray(a)) { content[q.id] = a; continue; }
    if (prop?.type === 'boolean') content[q.id] = a === 'true';
    else if (q.numeric) { const n = Number(a); if (Number.isFinite(n)) content[q.id] = n; }
    else content[q.id] = a;
  }
  return content;
}

File: rust/crates/acpira-host/src/acp/questions.rs
//! Structured questions: elicitation forms and Grok's question
//! request become one QuestionBlock; answers go back typed the way the schema declared them. The pending-card
//! bookkeeping lives in the session, under its lock

use std::collections::HashMap;

use serde_json::{Map, Value, json};

use acpira_shared::transcript::{Question, QuestionAnswers, QuestionKind, QuestionOption};

use crate::i18n::t;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct RawOption {
  pub label: String,
  pub description: Option<String>,
}

/// The agent's own tool input, the richer source for labels than the flattened form schema
#[derive(Debug, Clone, Default, PartialEq)]
pub struct RawQuestion {
  pub header: Option<String>,
  pub question: Option<String>,
  pub options: Option<Vec<RawOption>>,
}

const RAW_KEEP: usize = 8;

/// Remembered ask-user-question tool inputs, a handful is plenty
#[derive(Default)]
pub struct RawMemory {
  items: Vec<(String, Vec<RawQuestion>)>,
}

impl RawMemory {
  pub fn remember(&mut self, u: &Value) {
    let Some(questions) = u.get("rawInput").and_then(|r| r.get("questions")).and_then(Value::as_array) else { return };
    let id = u.get("toolCallId").and_then(Value::as_str).unwrap_or("").to_owned();
    let parsed: Vec<RawQuestion> = questions
      .iter()
      .filter_map(Value::as_object)
      .map(|item| RawQuestion {
        header: item.get("header").and_then(Value::as_str).map(str::to_owned),
        question: item.get("question").and_then(Value::as_str).map(str::to_owned),
        options: item.get("options").and_then(Value::as_array).map(|opts| {
          opts
            .iter()
            .filter_map(|o| {
              let label = o.get("label")?.as_str()?.to_owned();
              Some(RawOption { label, description: o.get("description").and_then(Value::as_str).map(str::to_owned) })
            })
            .collect()
        }),
      })
      .collect();
    self.items.retain(|(i, _)| *i != id);
    self.items.push((id, parsed));
    if self.items.len() > RAW_KEEP {
      let drop = self.items.len() - RAW_KEEP;
      self.items.drain(..drop);
    }
  }

  /// Devin sends no toolCallId with the form: fall back to the latest input with the same number of questions
  pub fn for_call(&self, tool_call_id: Option<&str>, count: usize) -> Option<Vec<RawQuestion>> {
    if let Some(id) = tool_call_id.filter(|x| !x.is_empty()) {
      return self.items.iter().find(|(i, _)| i == id).map(|(_, q)| q.clone());
    }
    self.items.iter().rev().find(|(_, q)| q.len() == count).map(|(_, q)| q.clone())
  }
}

fn nonblank(v: Option<&Value>) -> Option<String> {
  v.and_then(Value::as_str).filter(|s| !s.trim().is_empty()).map(str::to_owned)
}

fn plain(values: &[Value]) -> Vec<QuestionOption> {
  values
    .iter()
    .filter_map(|v| match v {
      Value::String(s) => Some(s.clone()),
      Value::Number(n) => Some(crate::json::pretty(&Value::Number(n.clone()))),
      _ => None,
    })
    .map(|s| QuestionOption { id: s.clone(), label: s, description: None })
    .collect()
}

fn titled(entries: &[Value], rq: Option<&RawQuestion>) -> Vec<QuestionOption> {
  entries
    .iter()
    .filter_map(|e| {
      let konst = e.get("const")?.as_str()?.to_owned();
      let ro = rq.and_then(|r| r.options.as_ref()).and_then(|opts| opts.iter().find(|o| o.label == konst));
      let label = ro.map(|o| o.label.clone()).or_else(|| nonblank(e.get("title"))).unwrap_or_else(|| konst.clone());
      let description = match ro {
        Some(o) => o.description.clone(),
        None => nonblank(e.get("description")),
      };
      Some(QuestionOption { id: konst, description: description.filter(|d| !d.is_empty() && *d != label), label })
    })
    .collect()
}

/// Codex's extra choice that stands for "my answer is in the note field" (codex-acp 1.13.0)
const CODEX_OTHER_OPTION: &str = "None of the above";

/// Adapter boilerplate in `message` that only says "a form follows": claude-agent-acp 0.81.0 for
/// several questions, codex-acp 1.13.0 for every request
const GENERIC_MESSAGES: [&str; 2] = ["Please answer the following questions.", "Codex needs your input to continue."];

/// A free-text property an adapter puts beside a choice question: Claude's per-question "Other" box
/// (`_meta._askUserQuestionCustomAnswer`), Codex's note field (`_meta.codex.role = user_note`)
struct Companion {
  key: String,
  /// The choice the parent must carry when the typed text is the answer
  sentinel: Option<&'static str>,
}

/// Companions by the key of the question they belong to
fn companions(props: &Map<String, Value>) -> HashMap<String, Companion> {
  let mut out = HashMap::new();
  for (key, prop) in props {
    let Some(meta) = prop.get("_meta") else { continue };
    let claude = meta.get("_askUserQuestionCustomAnswer").filter(|m| m.get("isCustomAnswer") == Some(&Value::Bool(true)));
    let codex = meta.get("codex").filter(|m| m.get("role").and_then(Value::as_str) == Some("user_note"));
    let (parent, sentinel) = match (claude, codex) {
      (Some(m), _) => (m.get("questionId"), None),
      (None, Some(m)) => (m.get("questionId"), Some(CODEX_OTHER_OPTION)),
      _ => continue,
    };
    let Some(parent) = parent.and_then(Value::as_str).filter(|p| *p != key && props.contains_key(*p)) else { continue };
    out.insert(parent.to_owned(), Companion { key: key.clone(), sentinel });
  }
  // A companion never has a companion of its own
  let keys: Vec<String> = out.values().map(|c| c.key.clone()).collect();
  out.retain(|parent, _| !keys.contains(parent));
  out
}

/// How many questions the form asks, companions folded into their questions
pub fn form_question_count(schema: &Value) -> usize {
  let Some(props) = schema.get("properties").and_then(Value::as_object) else { return 0 };
  props.len() - companions(props).len()
}

/// One question per form property, in schema order; companion properties become their question's "Other" row
pub fn form_questions(schema: &Value, message: &str, meta: Option<&Value>, raw: Option<&[RawQuestion]>) -> Vec<Question> {
  let empty = Map::new();
  let props = schema.get("properties").and_then(Value::as_object).unwrap_or(&empty);
  let companions = companions(props);
  let keys: Vec<&String> = props.keys().filter(|k| !companions.values().any(|c| c.key == **k)).collect();
  let required: Vec<&str> =
    schema.get("required").and_then(Value::as_array).map(|r| r.iter().filter_map(Value::as_str).collect()).unwrap_or_default();
  let allow_other = meta.and_then(|m| m.get("cognition.ai/allowOther")) == Some(&Value::Bool(true));
  let lines: Vec<&str> = message.split('\n').map(str::trim).filter(|l| !l.is_empty()).collect();
  let aligned = raw.filter(|r| r.len() == keys.len());
  let mut out = vec![];
  for (i, key) in keys.iter().enumerate() {
    let Some(prop) = props.get(*key).filter(|p| p.is_object()) else { continue };
    let rq = aligned.and_then(|a| a.get(i));
    let ptitle = nonblank(prop.get("title"));
    let pdesc = nonblank(prop.get("description"));
    let usable = rq.filter(|rq| {
      (ptitle.is_none() || rq.header.is_none() || ptitle == rq.header) && (pdesc.is_none() || rq.question.is_none() || pdesc == rq.question)
    });
    let title = ptitle.clone().or_else(|| usable.and_then(|u| u.header.clone()));
    let text = pdesc
      .clone()
      .or_else(|| usable.and_then(|u| u.question.clone()))
      .or_else(|| (lines.len() == keys.len()).then(|| lines[i].to_owned()))
      .or_else(|| if keys.len() == 1 { nonblank(Some(&Value::from(message))) } else { None })
      .or_else(|| title.clone())
      .unwrap_or_else(|| (*key).clone());
    let ptype = prop.get("type").and_then(Value::as_str);
    let mut q = Question {
      id: (*key).clone(),
      title: title.filter(|t| *t != text),
      text,
      kind: QuestionKind::Single,
      options: vec![],
      other: None,
      numeric: None,
      required: required.contains(&key.as_str()).then_some(true),
    };
    let items = if ptype == Some("array") { prop.get("items").filter(|i| i.is_object()) } else { None };
    if ptype == Some("boolean") {
      q.options = vec![
        QuestionOption { id: "true".into(), label: t("question.yes"), description: None },
        QuestionOption { id: "false".into(), label: t("question.no"), description: None },
      ];
    } else if let Some(items) = items {
      q.kind = QuestionKind::Multiple;
      q.options = if let Some(any) = items.get("anyOf").and_then(Value::as_array) {
        titled(any, usable)
      } else if let Some(e) = items.get("enum").and_then(Value::as_array) {
        plain(e)
      } else {
        vec![]
      };
    } else if let Some(one) = prop.get("oneOf").and_then(Value::as_array) {
      q.options = titled(one, usable);
    } else if let Some(e) = prop.get("enum").and_then(Value::as_array) {
      q.options = plain(e);
    } else {
      q.kind = QuestionKind::Text;
      if matches!(ptype, Some("number" | "integer")) {
        q.numeric = Some(true);
      }
    }
    let companion = companions.get(key.as_str());
    if let Some(sentinel) = companion.and_then(|c| c.sentinel) {
      q.options.retain(|o| o.id != sentinel);
    }
    let prop_other = prop.get("_meta").and_then(|m| m.get("cognition.ai/allowOther")) == Some(&Value::Bool(true));
    if q.kind != QuestionKind::Text && ptype != Some("boolean") && (allow_other || prop_other || companion.is_some()) {
      q.other = Some(true);
    }
    if q.kind != QuestionKind::Text && q.options.is_empty() {
      q.kind = QuestionKind::Text;
    }
    out.push(q);
  }
  out
}

/// Grok's question request → questions keyed by question text
pub fn grok_questions(req: &Value) -> Vec<Question> {
  req
    .get("questions")
    .and_then(Value::as_array)
    .into_iter()
    .flatten()
    .map(|q| {
      let text = q.get("question").and_then(Value::as_str).unwrap_or("").to_owned();
      Question {
        id: text.clone(),
        title: None,
        kind: if q.get("multiSelect") == Some(&Value::Bool(true)) { QuestionKind::Multiple } else { QuestionKind::Single },
        options: q
          .get("options")
          .and_then(Value::as_array)
          .into_iter()
          .flatten()
          .map(|o| {
            let label = o.get("label").and_then(Value::as_str).unwrap_or("").to_owned();
            let description = o.get("description").and_then(Value::as_str).filter(|d| !d.is_empty() && *d != label).map(str::to_owned);
            QuestionOption { id: label.clone(), label, description }
          })
          .collect(),
        text,
        other: Some(true),
        numeric: None,
        required: None,
      }
    })
    .collect()
}

/// The form's message earns a line of its own only when it says something the questions don't
pub fn spare_message(message: &str, questions: &[Question]) -> Option<String> {
  let m = message.trim();
  if m.is_empty() || GENERIC_MESSAGES.contains(&m) {
    return None;
  }
  let texts: Vec<&str> = questions.iter().map(|q| q.text.trim()).collect();
  let joined_lines: Vec<&str> = m.split('\n').map(str::trim).filter(|l| !l.is_empty()).collect();
  if texts.contains(&m) || texts.join("\n") == joined_lines.join("\n") {
    return None;
  }
  Some(m.to_owned())
}

/// Keep only answers to known questions with something in them
pub fn clean_answers(questions: &[Question], answers: &QuestionAnswers) -> QuestionAnswers {
  let mut out = QuestionAnswers::new();
  for q in questions {
    let Some(a) = answers.get(&q.id) else { continue };
    let list: Vec<String> = match a {
      Value::Array(items) => items.iter().filter_map(Value::as_str).map(str::to_owned).collect(),
      Value::String(s) => vec![s.clone()],
      _ => continue,
    };
    if q.kind == QuestionKind::Multiple {
      let list: Vec<Value> = list.iter().map(|s| s.trim()).filter(|s| !s.is_empty()).map(Value::from).collect();
      if !list.is_empty() {
        out.insert(q.id.clone(), Value::Array(list));
      }
    } else {
      let s = list.join(", ");
      let s = s.trim();
      if !s.is_empty() {
        out.insert(q.id.clone(), Value::from(s));
      }
    }
  }
  out
}

/// Answers typed per the schema: booleans and numbers converted, lists stay lists, everything else a string.
/// Typed text of a question with a companion goes into the companion, the picks stay on the question
pub fn form_content(schema: &Value, questions: &[Question], answers: &QuestionAnswers) -> Value {
  let empty = Map::new();
  let companions = companions(schema.get("properties").and_then(Value::as_object).unwrap_or(&empty));
  let mut content = Map::new();
  for q in questions {
    let Some(a) = answers.get(&q.id) else { continue };
    if let Some(c) = companions.get(&q.id).filter(|_| q.kind != QuestionKind::Text) {
      let items: Vec<&str> = match a {
        Value::Array(items) => items.iter().filter_map(Value::as_str).collect(),
        _ => a.as_str().into_iter().collect(),
      };
      let (mut picked, typed): (Vec<&str>, Vec<&str>) = items.into_iter().partition(|s| q.options.iter().any(|o| o.id == *s));
      if !typed.is_empty() {
        content.insert(c.key.clone(), Value::from(typed.join(", ")));
        picked.extend(c.sentinel);
      }
      if q.kind == QuestionKind::Multiple {
        if !picked.is_empty() {
          content.insert(q.id.clone(), Value::from(picked));
        }
      } else if let Some(p) = picked.first() {
        content.insert(q.id.clone(), Value::from(*p));
      }
      continue;
    }
    if a.is_array() {
      content.insert(q.id.clone(), a.clone());
      continue;
    }
    let s = a.as_str().unwrap_or("");
    let ptype = schema.get("properties").and_then(|p| p.get(&q.id)).and_then(|p| p.get("type")).and_then(Value::as_str);
    if ptype == Some("boolean") {
      content.insert(q.id.clone(), Value::Bool(s == "true"));
    } else if q.numeric == Some(true) {
      if let Some(n) = js_number(s) {
        content.insert(q.id.clone(), acpira_shared::num::js_number(n));
      }
    } else {
      content.insert(q.id.clone(), Value::from(s));
    }
  }
  Value::Object(content)
}

/// Number(s) for the shapes a numeric answer takes; None where JS would give NaN / Infinity
fn js_number(s: &str) -> Option<f64> {
  let t = s.trim();
  if t.is_empty() {
    return Some(0.0);
  }
  t.parse::<f64>().ok().filter(|n| n.is_finite())
}

pub fn grok_response(skip: bool, answers: &QuestionAnswers) -> Value {
  let empty = answers.is_empty();
  if skip || empty {
    if empty { json!({ "outcome": "skip_interview" }) } else { json!({ "outcome": "skip_interview", "partial_answers": answers }) }
  } else {
    json!({ "outcome": "accepted", "answers": answers })
  }
}
